experimental/collator/src/provider.rs

```diff
@@ -0,0 +1,276 @@
+#[derive(Debug, PartialEq, Clone)]
+#[cfg_attr(feature = "datagen", derive(serde::Serialize))]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+pub struct CollationJamoV1<'data> {
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub ce32s: ZeroVec<'data, u32>,
+}
+
+#[icu_provider::data_struct(CollationReorderingV1Marker = "collator/reord@1")]
+#[derive(Debug, PartialEq, Clone)]
+#[cfg_attr(feature = "datagen", derive(serde::Serialize))]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+pub struct CollationReorderingV1<'data> {
+    pub min_high_no_reorder: u32,
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub reorder_table: ZeroVec<'data, u8>, // len always 256
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub reorder_ranges: ZeroVec<'data, u32>,
+}
+
+impl<'data> CollationReorderingV1<'data> {
+    pub fn reorder(&self, primary: u32) -> u32 {
+        if let Some(b) = self.reorder_table.get((primary >> 24) as usize) {
+            if b != 0 || primary <= NO_CE_PRIMARY {
+                (u32::from(b) << 24) | (primary & 0x00FFFFFF)
+            } else {
+                self.reorder_ex(primary)
+            }
+        } else {
+            // GIGO case
+            debug_assert!(false);
+            primary
+        }
+    }
+
+    fn reorder_ex(&self, primary: u32) -> u32 {
+        if primary >= self.min_high_no_reorder {
+            return primary;
+        }
+        let q = primary | 0xFFFF;
+        for &range in self.reorder_ranges.as_ule_slice().iter() {
+            let r = u32::from_unaligned(range);
+            if q < r {
+                return primary.wrapping_add(r << 24);
+            }
+        }
+        // GIGO case
+        debug_assert!(false);
+        primary
+    }
+}
+
+#[icu_provider::data_struct(CollationMetadataV1Marker = "collator/meta@1")]
+#[derive(Debug, PartialEq, Clone, Copy)]
+#[cfg_attr(feature = "datagen", derive(serde::Serialize))]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+pub struct CollationMetadataV1 {
+    /// 8 highest bits for numeric primary.
+    /// The low 2 bits are the default strength - 1,
+    /// i.e. 0 for primary, 1 for secondary.
+    /// The rest for various flags, some of the space
+    /// being unused.
+    pub bits: u32,
+}
+
+impl CollationMetadataV1 {
+    const MAX_VARIABLE_MASK: u32 = 0b11;
+    const TAILORED_MASK: u32 = 1 << 3;
+    const TAILORED_DIACRITICS_MASK: u32 = 1 << 4;
+    const TAILORED_JAMO_MASK: u32 = 1 << 5;
+    const REORDERING_MASK: u32 = 1 << 6;
+    const LITHUANIAN_DOT_ABOVE_MASK: u32 = 1 << 7;
+    const BACWARD_SECOND_LEVEL_MASK: u32 = 1 << 8;
+    const ALTERNATE_SHIFTED_MASK: u32 = 1 << 9;
+    const CASE_FIRST_MASK: u32 = 1 << 10;
+    const UPPER_FIRST_MASK: u32 = 1 << 11;
+
+    #[inline(always)]
+    pub fn max_variable(&self) -> MaxVariable {
+        unsafe { core::mem::transmute((self.bits & CollationMetadataV1::MAX_VARIABLE_MASK) as u8) }
```

Comment:
Nit 1: Add a safety comment.

Nit 2 (optional): Move this code to instead be a constructor on `MaxVariable`.

Suggestion (optional): If you put `#[make_ule]` on your enum then you automatically get a safe u8-to-enum conversion function generated for you.

The only reason this is safe is because the enum has only 4 values (0, 1, 2, and 3), which are exactly those that your bitmask allows. If the enum had 3 or 5 values, this would not be safe.

Reply:
Filed as #1909